packages/datastore/src/sync/processors/mutation.ts

```
@@ -256,7 +267,7 @@ class MutationProcessor {
 						hasMore = (await this.outbox.peek(storage)) !== undefined;
 					});
 
-					this.observer.next!({
+					this.observer?.next?.({
```

It's entirely possible I'm being dense, but according to the PR description: "An `undefined` observer signals to the mutation processor not to immediately resume on mutation, but instead wait for an observer (the sync engine) to register itself and re-start() the mutation processor`. Wouldn't this mean that resume doesn't happen if the observer is `undefined`? And if so, why is it possible that the observer is possibly `undefined`, here?

---

I wrestled with this little part of the diff for longer than I care to admit, actually. TypeScript isn't convinced that `this.observer` always exists here, or that `.next` exists on the observer. So, with that said, I'm actually not 100% sure PR eliminates *all* cases where a mutation event goes missing. It does provide an incremental improvement to one edge I identified while trying to add tests.

Maybe it makes more sense to revert to bangs to at least get errors logged if someone hits that edge. I think we can shake out more edges after https://github.com/aws-amplify/amplify-js/pull/11131, which will give us a little more control to stress the protocol.

---

Hm, though we check that the observer isn't `undefined` at the beginning of `resume` (i.e. the status of `isReady`), it's possible that between there and here, the observer can now become `undefined`. I think I'm just most concerned about silent failures, and think that error logs would be better. Thoughts?

---

I think TypeScript is just confused that `this.isReady` actually asserts that `this.observer` is defined. Even if you place `this.isReady` directly before `this.observer.next` you will still get the same error.

```
if (this.isReady()) {
  // still thinks that this.observer could be undefined
  this.observer.next ...
```

You could use assertion predicates to give TypeScript enough information to know it would be defined in this code path. But it looks like it isn't well supported on private class properties and would require some hacks.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

https://stackoverflow.com/questions/74264474/in-typescript-how-do-i-assert-the-value-of-a-class-variable-using-a-method